### PR TITLE
Fix failing provider tests

### PR DESCRIPTION
**Pull request by timrae**
The provider was using checkCallingPermission() which only checks if another process has the permission. Since the unit tests originate from the same process as AnkiDroid, it needs to be checkCallingOrSelfPermission(). I also added a task that automatically grants permission to the emulator over adb for the two permissions that we need to run the tests. The adb calls will be rejected below API 23 (our tests currently run on an API22 emulator), so they are mainly for when we eventually upgrade.

@marcardar
FYI

### CODE REVIEW

**Comment by marcardar** [comment]
Hmm interesting. I'm a bit nervous about the new checkCallingOrSelfPermission() because that will mean that if (for some reason) AnkiDroid is granted that permission then all apps (including client apps without that permission) will be able to perform update/delete.  The issue here is that the provider tests should not be running in the same process as AnkiDroid.

**Comment by timrae** [comment]
> The issue here is that the provider tests should not be running in the same process as AnkiDroid.

The provider is part of com.ichi2.anki, so this is normal. I think you're getting confused with what would happen if the checks were called from the API?

**Comment by marcardar** [comment]
The provider is designed to be called from another process. Therefore if you are testing the provider, then you should be running those tests from another process.

**Comment by timrae** [comment]
> if (for some reason) AnkiDroid is granted that permission then all apps (including client apps without that permission) will be able to perform update/delete.

AnkiDroid will always have that permission... Does that mean this PR will make the checks not do anything?!

**Comment by marcardar** [comment]
> AnkiDroid will always have that permission...

Are you sure? I thought it was only declared, not specified.

**Comment by timrae** [comment]
Yes I tested this... Declaring the permission seems to grant is as well, which makes sense I guess.

**Comment by marcardar** [comment]
> Yes I tested this... Declaring the permission seems to grant is as well, which makes sense I guess.

I'm quite surprised. And I wonder why does it fail the checkCallingPermission() if the tests (running in AnkiDroid process) already have that permission?

**Comment by timrae** [comment]
> And I wonder why does it fail the checkCallingPermission() if the tests (running in AnkiDroid process) already have that permission?

checkCallingPermission() returns -1 in this case. The [documentation](http://developer.android.com/reference/android/content/Context.html#checkCallingPermission(java.lang.String)) says:

> if you are not currently processing an IPC, this function will always fail. This is done to protect against accidentally leaking permissions; you can use checkCallingOrSelfPermission(String) to avoid this protection.

**Comment by marcardar** [comment]
Ah yes, I forgot about that. So yes, it looks like AnkiDroid has its own permission. Still, the test case is not testing the permissions part (for IPC).  Maybe that can be done another day?

File: AnkiDroid/travis.gradle [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    commandLine "$adb shell pm grant com.ichi2.anki android.permission.WRITE_EXTERNAL_STORAGE".split(' ')
+}
+task setProviderPermission(type: Exec, dependsOn: 'installDebug') {
+    commandLine "$adb shell pm grant com.ichi2.anki com.ichi2.anki.permission.READ_WRITE_DATABASE".split(' ')
+}
+
+tasks.whenTaskAdded { task ->
+    if (task.name.startsWith('connectedDebugAndroidTest')) {
+        task.dependsOn setStoragePermission
+        task.dependsOn setProviderPermission
```
Comment (marcardar) [inline AnkiDroid/travis.gradle]: Can you please confirm it fails when you remove this line?

**Comment by timrae** [comment]
I haven't tested it yet but from what I read this change will return yes for checkCallingOrSelfPermission(). Maybe a better way to address this issue is to make a new provider that uses proper static permissions, and have the API call that on SDK 23 and above.

**Comment by marcardar** [comment]
Yeah, that would work in the long-term, but we need the runtime permission check in the meantime, and that should work on all API levels.  How about a hack solution - let it pass if the calling package is null on API levels 19+?  Because that means it's not an IPC call.

**Comment by marcardar** [comment]
Whoops - let me rephrase that...

**Comment by timrae** [comment]
> we need the runtime permission check in the meantime

We can just switch the tests to use the new provider, then they won't fail.

**Comment by marcardar** [comment]
and run the tests on 23+ emulator?

**Comment by timrae** [comment]
> and run the tests on 23+ emulator?

That is not necessary since the emulator doesn't suffer from install order bug

**Comment by marcardar** [comment]
I just checked, and getCallingPackage() returns "com.ichi2.anki" as expected. So we could just do:

```
if (getContext().checkCallingPermission(FlashCardsContract.READ_WRITE_PERMISSION) != PackageManager.PERMISSION_GRANTED && !isAnkiCalling()) {
...

private void isAnkiCallingFromKitkat() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
            return false;
        } else {
            return "com.ichi2.anki".equals(getCallingPackage());
        }
}
```

**Comment by marcardar** [comment]
```
private void isUpdateDeleteAllowed() {
    if (getContext().checkCallingPermission(FlashCardsContract.READ_WRITE_PERMISSION) == PackageManager.PERMISSION_GRANTED) {
        return true;
    }
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
        return false; // we can't know if it's this app or not
    } else {
        return getPackageName().equals(getCallingPackage());
    }
}
```
